chat.go

```diff
@@ -0,0 +1,190 @@
+// Package chat implements a chat server & client
+package chat
+
+import (
+	"bufio"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+// Server represents a chat server
+type Server struct {
+	mutex         sync.Mutex
+	listener      net.Listener
+	running       bool
+	host          string
+	port          int
+	Logger        logger // with standard logger can be extended with logrus
+	ListenAddress string
+}
+
+// Logger enables a customization of the log function
+type logger interface {
+	Println(v ...interface{})
+}
+
+// StandardLogger defines a standard logger that implements the Logger interface
+type StandardLogger struct {
+	timeFormat string
+}
+
+// Client represents a chat client
+type Client struct {
+	connection net.Conn
+}
+
+// StartServer returns a pointer to a running server on localhost and random port
+func StartServer(options ...func(*Server)) (*Server, error) {
+
+	rand.Seed(time.Now().UnixNano())
+	p := 8080 + rand.Intn(100) // Add used port detection
+
+	logger := NewStandardLogger(time.RFC3339)
+
+	s := &Server{
+		running: true,
+		host:    "localhost",
+		port:    p,
+		Logger:  logger,
+	}
+
+	for _, option := range options {
+		option(s)
+	}
+
+	s.ListenAddress = fmt.Sprintf(s.host+":%d", s.port)
+
+	ln, err := net.Listen("tcp", s.ListenAddress)
+	if err != nil {
+		return &Server{}, err
+	}
+
+	s.listener = ln
+
+	go s.Run()
+	return s, nil
+}
+
+// Run implements the logic handling connections
+func (s *Server) Run() {
+	conn, err := s.listener.Accept()
+	if err != nil {
+		s.Logger.Println(fmt.Sprintf("connection failed: %v", err))
+		return
+	}
+	if !s.Running() {
+		conn.Close() // Not sure if it is still useful as listener.close closes all connections
+		return
+	}
+	_, err = conn.Write([]byte("Welcome to ChatRoom !\n"))
+	if err != nil {
+		s.Logger.Println(fmt.Sprintf("sending message failed: %v", err))
+	}
+
+	r := bufio.NewReader(conn)
+
+	for s.Running() {
+		m, err := r.ReadString('\n')
+		if err != nil {
+			s.Logger.Println(fmt.Sprintf("receiving message failed: %v", err))
+		}
+		// fmt.Printf("server: message received %q", m)
+		_, err = fmt.Fprintf(conn, m)
+		if err != nil {
+			s.Logger.Println(fmt.Sprintf("sending message failed: %v", err))
+		}
+		// fmt.Printf("server: message sent %q", m)
+	}
+}
+
+// Running indicates if the server can accept connections
+func (s *Server) Running() bool {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.running
+}
+
+// Stop stops a running server
+func (s *Server) Stop() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.running = false
+	s.listener.Close()
+}
+
+// ListenAndServe blocks while the server is running
+func (s *Server) ListenAndServe() {
+	for s.Running() {
+		time.Sleep(5 * time.Second)
```

Comment:
While studying for this done channel implementation and the port customization, I'm more and more thinking about a massive refactoring of my Server methods.
I think it would be easier for the user if it was closer from the standard library implementation of a server.
I will change StartServer to NewServer, move the logic from run to ListenServe and rename Stop to Close.
In my opinion, it would make more sense.